pkg/certificate/manager.go

```diff
@@ -0,0 +1,146 @@
+
+// NewManager creates a new CertManager with the passed CA and CA Private Key
+func NewManager(
+	ca *Certificate,
+	client client,
+	serviceCertValidityDuration time.Duration,
+	msgBroker *messaging.Broker) (*manager, error) {
+	if ca == nil {
+		return nil, errNoIssuingCA
+	}
+
+	m := &manager{
+		// The root certificate signing all newly issued certificates
+		ca:                          ca,
+		client:                      client,
+		serviceCertValidityDuration: serviceCertValidityDuration,
+		msgBroker:                   msgBroker,
+	}
+
+	// TODO(steeling) start the cert rotation here.
+
+	return m, nil
+}
+
+func (m *manager) deleteFromCache(cn CommonName) {
+	m.cache.Delete(cn)
+}
+
+func (m *manager) getFromCache(cn CommonName) *Certificate {
+	if certInterface, exists := m.cache.Load(cn); exists {
+		cert := certInterface.(*Certificate)
+		log.Trace().Msgf("Certificate found in cache SerialNumber=%s", cert.GetSerialNumber())
+		if cert.ShouldRotate() {
+			log.Trace().Msgf("Certificate found in cache but has expired SerialNumber=%s", cert.GetSerialNumber())
+			return nil
+		}
+		return cert
+	}
+	return nil
+}
+
+// IssueCertificate implements Manager and returns a newly issued
+func (m *manager) IssueCertificate(cn CommonName, validityPeriod time.Duration) (*Certificate, error) {
+	start := time.Now()
+
+	if cert := m.getFromCache(cn); cert != nil {
+		return cert, nil
+	}
+
+	cert, err := m.client.IssueCertificate(cn, validityPeriod)
+	if err != nil {
+		return cert, err
+	}
+
+	m.cache.Store(cn, cert)
+
+	log.Trace().Msgf("It took %+v to issue certificate with SerialNumber=%s", time.Since(start), cert.GetSerialNumber())
+
+	return cert, nil
+}
+
+// ReleaseCertificate is called when a cert will no longer be needed and should be removed from the system.
+func (m *manager) ReleaseCertificate(cn CommonName) {
+	log.Trace().Msgf("Releasing certificate %s", cn)
+	m.deleteFromCache(cn)
+}
+
+// GetCertificate returns a certificate given its Common Name (CN)
+func (m *manager) GetCertificate(cn CommonName) (*Certificate, error) {
+	if cert := m.getFromCache(cn); cert != nil {
+		return cert, nil
+	}
+	return nil, errCertNotFound
+}
+
+// RotateCertificate implements Manager and rotates an existing
+func (m *manager) RotateCertificate(cn CommonName) (*Certificate, error) {
+	start := time.Now()
+
+	oldObj, ok := m.cache.Load(cn)
+	if !ok {
+		return nil, errors.Errorf("Old certificate does not exist for CN=%s", cn)
+	}
+
+	oldCert, ok := oldObj.(*Certificate)
+	if !ok {
+		return nil, errors.Errorf("unexpected type %T for old certificate does not exist for CN=%s", oldCert, cn)
+	}
+
+	newCert, err := m.IssueCertificate(cn, m.serviceCertValidityDuration)
+	if err != nil {
+		return nil, err
+	}
+
+	m.cache.Store(cn, newCert)
+
+	m.msgBroker.GetCertPubSub().Pub(events.PubSubMessage{
+		Kind:   announcements.CertificateRotated,
+		NewObj: newCert,
+		OldObj: oldCert,
+	}, announcements.CertificateRotated.String())
```

Comment:
Since we're going down to a single place where we do this, I'm think we can be less concerned about introducing bugs than the current state where there was one instance per cert provider.

I think you highlight a good refactoring opportunity for the msgbroker though.

It may make more sense to provide more targeted methods, ie: `PublishCertRotation`, or provide generic methods that type switch based on the incoming type. Haven't looked at this code too much though